pallets/randomness/src/vrf.rs

```diff
@@ -0,0 +1,126 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+//! VRF logic
+use crate::{Config, CurrentVrfInput, GetVrfInput, LocalVrfOutput, RandomnessResults, RequestType};
+use frame_support::{pallet_prelude::Weight, traits::Get};
+use nimbus_primitives::{NimbusId, NIMBUS_ENGINE_ID};
+use parity_scale_codec::{Decode, Encode};
+use scale_info::TypeInfo;
+#[cfg(feature = "std")]
+use serde::{Deserialize, Serialize};
+pub use session_keys_primitives::make_transcript;
+use session_keys_primitives::{KeysLookup, PreDigest, VrfId, VRF_ENGINE_ID, VRF_INOUT_CONTEXT};
+use sp_consensus_vrf::schnorrkel;
+use sp_core::crypto::ByteArray;
+use sp_runtime::RuntimeDebug;
+
+/// VRF output
+type Randomness = schnorrkel::Randomness;
+
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Default, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+/// VRF inputs from the relay chain
+/// Both inputs are expected to change every block
+pub struct VrfInput<SlotNumber, RelayHash> {
+	/// Relay block slot number
+	pub slot_number: SlotNumber,
+	/// Relay block storage root
+	pub storage_root: RelayHash,
+}
+
+/// Set vrf input in storage and log warning if either of the values did NOT change
+/// Called in previous block's `on_finalize`
+pub(crate) fn set_input<T: Config>() {
+	let input = T::VrfInputGetter::get_vrf_input();
+	if let Some(last_vrf_input) = <CurrentVrfInput<T>>::take() {
+		// logs if input uniqueness assumptions are violated (no reuse of vrf inputs)
+		if last_vrf_input.storage_root == input.storage_root
+			|| last_vrf_input.slot_number == input.slot_number
+		{
+			log::warn!(
+				"VRF on_initialize: storage root or slot number did not change between \
+			current and last block. Nimbus would've panicked if slot number did not change \
+			so probably storage root did not change."
+			);
+		}
+	}
+	<CurrentVrfInput<T>>::put(input);
+}
+
+/// Returns weight consumed in `on_initialize`
+pub(crate) fn set_output<T: Config>() -> Weight {
+	let input = <CurrentVrfInput<T>>::get().expect("VrfInput must be set to verify VrfOutput");
+	let mut block_author_vrf_id: Option<VrfId> = None;
+	let PreDigest {
+		vrf_output,
+		vrf_proof,
+	} = <frame_system::Pallet<T>>::digest()
+		.logs
+		.iter()
+		.filter_map(|s| s.as_pre_runtime())
+		.filter_map(|(id, mut data)| {
+			if id == VRF_ENGINE_ID {
+				PreDigest::decode(&mut data).ok()
+			} else {
+				if id == NIMBUS_ENGINE_ID {
+					let nimbus_id = NimbusId::decode(&mut data)
+						.expect("NimbusId encoded in pre-runtime digest must be valid");
+
+					block_author_vrf_id = Some(
+						T::VrfKeyLookup::lookup_keys(&nimbus_id)
+							.expect("No VRF Key Mapped to this NimbusId"),
```

Comment:
Is this check optional? Seems like it would get ignored if there was somehow no `id == NIMBUS_ENGINE_ID`

Reply:
There is an unwrap of block_author_vrf_id below the iterator which panics if the value is None after going through all digests.